### DISCUSSION TITLE
"Error setting curl" when using curlftpfs

### DISCUSSION THREAD

**Discussion by FelixJochems** [curl/curl#14299] [Q&A] [answered]
Anyone else having this issue atm? I use curlftpfs to connect to a local ftp share and even though it's technically an issue with curlftpfs, it hasn't been updated in over 10 years making me think it is something with curl 8.9.0.
Downgrading to 8.8.0 solves the issue for now.

command is used in the following format:
```
curlftpfs USER:PASSWD@HOST /mnt/ftp
```

**Comment by bagder** [comment]
You need to tell us a lot more. curlftpfs is not a curl project, we don't know how that works. What does *Error setting curl* mean?

Also, did you build both your libcurls exactly the same?

**Comment by FelixJochems** [comment]
I thought this issue would be a lot more common. Seeing that nobody has the same issue it's very likely to be local to my machine, I'll investigate/debug and come back with more detailed info if I find it.

Thank you for the quick response!

**Comment by LaborMatthias** [comment]
Using gentoo, getting a prefix of a warning "Error setting curl:" without the real warning :-/
Probably due to an update of either curl to 8.8.0-r1 or curlftpfs to 0.9.2-r5
So I cannot mount an ftp server with one PC, while I still can with another (without these updates, gotta check what versions it uses...)

**Comment by user202729** [comment]
Just for a datapoint. I'm using Arch Linux, just start getting the error. The error message is just

```
Error setting curl: 
```

(in other words, the returned error message is **empty**.)

Downgrading to curl 8.8.0 works.

**Comment by uninsane** [comment]
specifically, curl's returning an error for `curl_easy_setopt(handle, CURLOPT_INTERFACE, NULL)` in 8.9.0, but not 8.8.0. for curlftpfs, this can be resolved by passing the `interface` mount option, like `curlftpfs USER:PASSWD@HOST /mnt/ftp -o interface=0.0.0.0`.

**Reply by bagder** [reply]
> curl's returning an error for curl_easy_setopt(handle, CURLOPT_INTERFACE, NULL) in 8.9.0

That sounds like a bug.

**Reply by bagder** [reply]
See  #14629

**Comment by LaborMatthias** [comment]
I tried curlftpfs ... -o interface=0.0.0.0 as well as with the name of the interface - without any change.
But I also tried to use curl directly on that server with this command which successfully listed the directory (no -o interface given):
curl -l ftp://user:pwd@host/media/hdd/
and I also downloaded a file.
Then I tried to strace -f the curlftpfs with about no clue found: The result did only show the following hits for the string "curl" after the first line:
openat(AT_FDCWD, "/usr/lib64/libcurl.so.4", O_RDONLY|O_CLOEXEC) = 3
write(2, "Error setting curl: \n", 21Error setting curl: 
So I believe in my situation curl_easy_setopt was never called.

**Reply by uninsane** [reply]
`strace -f` is only showing you syscalls, it's not going to trace the function calls between curlftpfs and curl. but you could get this via gdb:

```
$ gdb curlftpfs
(gdb) break exit
(gdb) run -o interface=0.0.0.0
...
wait until it hits the breakpoint
(gdb) backtrace
this should show you where in curlftpfs it died
```

if you can build for debug mode, that will get you a more useful trace. if you can build from source, [this](https://git.uninsane.org/colin/curlftpfs/commit/a02006d8ecbcdc5d71102ccfc5e6fa74a82e66a5.patch) patch will tell you more info even without gdb.

**Comment by beelze** [comment]
same issue. `curlftpfs-0.9.2` shows only `Error setting curl: ` with `curl-8.8.0`. Reverting curl to 8.7.1 solves an issue.

**Comment by bagder** [comment, marked as answer]
This bug is fixed in current master, set to be released in curl 8.10.0 on September 11, 2024.

**Reply by d3jv** [reply]
I'm still seeing this bug on 8.10.1

**Reply by LaborMatthias** [reply]
Me too, but using curl by itself does not have the problem as far as I tested it. You might want to open an issue about curlftpfs and post the link here

**Reply by dymn** [reply]
I'm still seeing this bug on 8.11.0.

**Reply by ghost** [reply]
Still the same issue for me on Void Linux, `curl` 8.12.1 and `curlftpfs` 0.9.2, tried [this fork ](https://github.com/ikn/curlftpfs-ng), but still see the error. The behavior is pretty much the same as described here: [link1](https://www.linuxquestions.org/questions/slackware-14/problem-with-curlftpfs-4175740087/) [link2](https://bbs.archlinux.org/viewtopic.php?id=298351) [link3](https://bugs.launchpad.net/ubuntu/+source/curlftpfs/+bug/2095153)

Everything used to work fine, so I suspect the issue might be in `curl` itself

**Reply by bagder** [reply]
It would be more useful if you told us which libcurl API call that returns error. I presume it is similar to the one fixed before. While the symptom looks the same, we know it is not the same bug if you run a fixed libcurl.

**Reply by ghost** [reply]
Yes, the error is different. If I use an official `curlftpfs` from sourceforge, I see that it fails during `curl_easy_setopt_or_die(easy, CURLOPT_NOBODY, ftpfs.safe_nobody)`, but [the patched version](https://github.com/ikn/curlftpfs-ng) shows a different function call `curl_easy_setopt_or_die(easy, CURLOPT_IPRESOLVE, ftpfs.ip_version)`.

Also, passing an option like `-o ipv4` doesn't eliminate the error

<details> 
  <summary>Sourceforge version</summary>
  
```
$ gdb --args curlftpfs -f -v -o debug,ftpfs_debug=3,curldebug=1 192.168.1.123:1234 /home/user/ftp/


GNU gdb (GDB) 16.2
Copyright (C) 2024 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "x86_64-unknown-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<https://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
    <http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from curlftpfs...
Reading symbols from /usr/lib/debug/usr/bin/curlftpfs...
(gdb) break main
Breakpoint 1 at 0x3560: file ftpfs.c, line 1726.
(gdb) run
Starting program: /usr/bin/curlftpfs -f -v -o debug,ftpfs_debug=3,curldebug=1 192.168.1.123:1234 /home/user/ftp/
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/usr/lib64/libthread_db.so.1".


Breakpoint 1, main (argc=7, argv=0x7fffffffc498) at ftpfs.c:1726
1726      struct fuse_args args = FUSE_ARGS_INIT(argc, argv);
(gdb) n
1736      // Set some default values
(gdb)
1728      CURL* easy;
(gdb)
1734      memset(&ftpfs, 0, sizeof(ftpfs));
(gdb)
1736      // Set some default values
(gdb)

1739      ftpfs.blksize = 4096;
(gdb)
1740      ftpfs.disable_epsv = 1;
(gdb)
1743
(gdb)
1744      if (fuse_opt_parse(&args, &ftpfs, ftpfs_opts, ftpfs_opt_proc) == -1)
(gdb)
1746
(gdb)
1749        fprintf(stderr, "see `%s -h' for usage\n", argv[0]);
(gdb)
1755      }
(gdb)
1756
(gdb)
1759      }
(gdb)
1763        fprintf(stderr, "Error initializing libcurl\n");
(gdb)
1764        exit(1);
(gdb)
1769        exit(1);
(gdb)
1770
(gdb)
1773
(gdb)
1774      if (ftpfs.transform_symlinks && !ftpfs.mountpoint) {
(gdb)
1776        exit(1);
(gdb)
1781        ftpfs.symlink_prefix_len = strlen(ftpfs.symlink_prefix);
(gdb)
1789      curl_easy_setopt_or_die(easy, CURLOPT_NOBODY, ftpfs.safe_nobody);
(gdb)
Error setting curl:
[Inferior 1 (process 8345) exited with code 01]
```

</details> 

<details> 
  <summary>Patched version</summary>
  
```
$ gdb --args ./curlftpfs -f -v -o debug,ftpfs_debug=3,curldebug=1 192.168.1.123:1234 /home/user/ftp/
GNU gdb (GDB) 16.2
Copyright (C) 2024 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "x86_64-unknown-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<https://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
    <http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from ./curlftpfs...
(gdb) break set_common_curl_stuff
Breakpoint 1 at 0x3ce0: file ftpfs.c, line 1521.
(gdb) run
Starting program: /home/user/src/curlftpfs-ng/curlftpfs -f -v -o debug,ftpfs_debug=3,curldebug=1 192.168.1.123:1234 /home/user/ftp/
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/usr/lib64/libthread_db.so.1".

Breakpoint 1, set_common_curl_stuff (easy=easy@entry=0x555555585250) at ftpfs.c:1521
1521      curl_easy_setopt_or_die(easy, CURLOPT_WRITEFUNCTION, read_data);
(gdb) next
1522      curl_easy_setopt_or_die(easy, CURLOPT_READFUNCTION, write_data);
(gdb) next
1523      curl_easy_setopt_or_die(easy, CURLOPT_ERRORBUFFER, error_buf);
(gdb) next
1524      curl_easy_setopt_or_die(easy, CURLOPT_URL, ftpfs.host);
(gdb) n
1525      curl_easy_setopt_or_die(easy, CURLOPT_NETRC, CURL_NETRC_OPTIONAL);
(gdb) n
1526      curl_easy_setopt_or_die(easy, CURLOPT_NOSIGNAL, 1);
(gdb) n
1527      curl_easy_setopt_or_die(easy, CURLOPT_CUSTOMREQUEST, "LIST -a");
(gdb) n
1529      if (ftpfs.custom_list) {
(gdb) n
1533      if (ftpfs.tryutf8) {
(gdb) n
1546      if (ftpfs.verbose) {
(gdb) n
1547        curl_easy_setopt_or_die(easy, CURLOPT_VERBOSE, TRUE);
(gdb) n
1550      if (ftpfs.disable_epsv) {
(gdb) n
1551        curl_easy_setopt_or_die(easy, CURLOPT_FTP_USE_EPSV, FALSE);
(gdb) n
1554      if (ftpfs.skip_pasv_ip) {
(gdb) n
1558      if (ftpfs.ftp_port) {
(gdb) n
1562      if (ftpfs.disable_eprt) {
(gdb) n
1566      if (ftpfs.ftp_method) {
(gdb) n
1571      if (ftpfs.tcp_nodelay) {
(gdb) n
1576      curl_easy_setopt_or_die(easy, CURLOPT_CONNECTTIMEOUT, ftpfs.connect_timeout);
(gdb) n
1581      if (ftpfs.use_ssl > CURLFTPSSL_TRY &&
(gdb) n
1598      curl_easy_setopt_or_die(easy, CURLOPT_FTP_SSL, ftpfs.use_ssl);
(gdb) n
1600      curl_easy_setopt_or_die(easy, CURLOPT_SSLCERT, ftpfs.cert);
(gdb) n
1601      curl_easy_setopt_or_die(easy, CURLOPT_SSLCERTTYPE, ftpfs.cert_type);
(gdb) n
1602      curl_easy_setopt_or_die(easy, CURLOPT_SSLKEY, ftpfs.key);
(gdb) n
1603      curl_easy_setopt_or_die(easy, CURLOPT_SSLKEYTYPE, ftpfs.key_type);
(gdb) n
1604      curl_easy_setopt_or_die(easy, CURLOPT_SSLKEYPASSWD, ftpfs.key_password);
(gdb) n
1606      if (ftpfs.engine) {
(gdb) n
1611      curl_easy_setopt_or_die(easy, CURLOPT_SSL_VERIFYPEER, TRUE);
(gdb) n
1612      if (ftpfs.no_verify_peer) {
(gdb) n
1616      if (ftpfs.cacert || ftpfs.capath) {
(gdb) n
1620        if (ftpfs.capath) {
(gdb) n
1625      if (ftpfs.ciphers) {
(gdb) n
1629      if (ftpfs.no_verify_hostname) {
(gdb) n
1633      curl_easy_setopt_or_die(easy, CURLOPT_INTERFACE, ftpfs.interface);
(gdb) n
1634      curl_easy_setopt_or_die(easy, CURLOPT_KRB4LEVEL, ftpfs.krb4);
(gdb) n
1664      curl_easy_setopt_or_die(easy, CURLOPT_IPRESOLVE, ftpfs.ip_version);
(gdb) n
Error setting curl:
[Inferior 1 (process 19924) exited with code 01]
</details>

**Reply by bagder** [reply]
> it fails during curl_easy_setopt_or_die(easy, CURLOPT_NOBODY, ftpfs.safe_nobody),

I don't see how setting `CURLOPT_NOBODY` can even return an error. But I can tell you one thing: `ftpfs.safe_nobody` is an `int` and *setopt() requires a `long` so this is incorrect use of the function. You should probably either add a typecast in the invocation or change the type of the struct field.

`ftpfs.ip_version` has the exact same problem. libcurl then probably reads junk off the stack for the other 32 bits and that's why it returns an error because it gets a bogus setting value there.

These are both bugs in the curlftpfs code. I checked curlftpfs-ng source code, but I bet it looks similar in the original code.

**Comment by ChrisVine** [comment]
This bug seems to have re-emerged in curl-8.17.0.  Reverting to curl-8.16.0 resolves the issue.  With curl-8.17.0 mounting via curlftpfs fails with 'Error setting curl:' .

**Reply by bagder** [reply]
I don't think it's the same issue even if the symptom might look the same.

**Reply by SuicideShow** [reply]
Got the same problem, going back to curl-8.16 helped.

**Reply by bagder** [reply]
And until someone figures out what option it tries to set (and to what) that libcurl rejects, the problem will not be fixed.

**Reply by LawnGnome** [reply]
This is caused by the removal of Kerberos FTP support in 8.17.0; curlftpfs unconditionally sets `CURLOPT_KRB4LEVEL`, and exits if `curl_easy_setopt` returns an error, which it (reasonably) now does.

I posted a patch to https://sourceforge.net/p/curlftpfs/patches/16/ to fix this on the curlftpfs side.
